okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java

```diff
@@ -228,89 +173,36 @@ private boolean checkTokenCache(String tenant, HttpServerRequest req, String pat
     final String id = req.getHeader(XOkapiHeaders.MODULE_ID);
-    logger.debug("getModulesForRequest: Matching {} {}", req.method(), req.uri());
-
+    List<ModuleInstance> mods = null;
+    try {
+      mods = moduleCache.lookup(req.uri(), req.method(), id);
+    } catch (IllegalArgumentException e) {
+      pc.responseError(500, e.getMessage());
+      return null;
+    }
     boolean skipAuth = false;
-    Timer.Sample sampleLoopEnabledModules = MetricsHelper.getTimerSample();
-
-    for (ModuleDescriptor md : enabledModules) {
-      logger.debug("getModulesForRequest:  looking at {}", md.getId());
-      List<RoutingEntry> rr = null;
-      if (id == null) {
-        rr = md.getProxyRoutingEntries();
-      } else if (id.equals(md.getId())) {
-        rr = md.getMultiRoutingEntries();
-      }
-      if (rr != null) {
-        Timer.Sample sampleLoopRoutingEntries = MetricsHelper.getTimerSample();
-        for (RoutingEntry re : rr) {
-          if (match(re, req)) {
-            ModuleInstance mi = new ModuleInstance(md, re, req.uri(), req.method(), true);
-
-            skipAuth = checkTokenCache(pc.getTenant(), req, re.getPathPattern(), mi);
-            mods.add(mi);
-            logger.debug("getModulesForRequest:  Added {} {} {} {} / {}", md.getId(),
-                re.getPathPattern(), re.getPath(), re.getPhase(), re.getLevel());
-            break;
-          }
-        }
-        MetricsHelper.recordCodeExecutionTime(sampleLoopRoutingEntries,
-            name + ".loopRoutingEntries");
+    for (ModuleInstance mi : mods) {
+      if (mi.isHandler()) {
+        RoutingEntry re = mi.getRoutingEntry();
+        skipAuth = checkTokenCache(pc.getTenant(), req, re.getPathPattern(), mi);
+        logger.debug("getModulesForRequest:  Added {} {} {} {} / {}",
+            mi.getModuleDescriptor().getId(),
+            re.getPathPattern(), re.getPath(), re.getPhase(), re.getLevel());
       }
-      rr = md.getFilterRoutingEntries();
-      Timer.Sample sampleLoopFilterEntries = MetricsHelper.getTimerSample();
-      for (RoutingEntry re : rr) {
-        if (match(re, req)) {
-          ModuleInstance mi = new ModuleInstance(md, re, req.uri(), req.method(), false);
-          mi.setAuthToken(req.headers().get(XOkapiHeaders.TOKEN));
-          mods.add(mi);
-          if (!resolveRedirects(pc, mods, re, enabledModules, "", req.uri())) {
-            MetricsHelper.recordCodeExecutionTime(sampleLoopFilterEntries,
-                name + ".loopFilterEntries");
-            MetricsHelper.recordCodeExecutionTime(sampleLoopEnabledModules,
-                name + ".loopEnabledModules");
-            MetricsHelper.recordCodeExecutionTime(sample, name);
-            return null;
-          }
-          logger.debug("getModulesForRequest:  Added {} {} {} {} / {}", md.getId(),
-              re.getPathPattern(), re.getPath(), re.getPhase(), re.getLevel());
-        }
+    }
+    for (ModuleInstance mi : mods) {
```

Comment:
Can this loop be combined with above one?

Reply:
Yes. Will do.